lib/msf/core/exploit/remote/java/http/class_loader.rb

```diff
@@ -54,22 +59,25 @@ def on_request_uri(cli, request)
       return send_response(cli, '')
     end
 
-    case request.uri
+    case resource
     # Stage 1
-    when "/#{class_name}.class"
+    when "#{class_name}.class"
       vprint_good('Sending constructor class')
       # This contains the constructor that will call our JavaPayload
       res = constructor_class
     # Stage 2
-    when '/metasploit/Payload.class'
+    when 'metasploit/Payload.class'
       vprint_good('Sending payload class')
       # This is our JavaPayload as a compiled class
       res = MetasploitPayloads.read('java/metasploit/Payload.class')
     # Stage 3
-    when '/metasploit.dat'
+    when 'metasploit.dat'
       vprint_good('Sending payload config')
       # This tells the target how to address the payload; this is the magic!
       res = payload_instance.stager_config
+    # (Optional) Stage 4 data for unstaged payloads such as java/shell_reverse_tcp
+    when /^javapayload\/stage\/(Shell|Stage|StreamForwarder)\.class$/
```

Comment:
```suggestion
    when /^javapayload\/stage\/(?:Shell|Stage|StreamForwarder)\.class$/
```

no need to have the group be a matching one.